File: Classes/server/TKReporter.m
//
//  TKReporter.m
//  TripGo
//
//  Created by Adrian Schoenig on 8/07/2015.
//  Copyright (c) 2015 SkedGo Pty Ltd. All rights reserved.
//

#import "TKReporter.h"

#import "TKTripKit.h"

@implementation TKReporter

+ (void)reportPlannedTrip:(Trip *)trip
                 userInfo:(NSDictionary *)userInfo
        completionHandler:(void(^)(BOOL success))completeHandler
{
  NSParameterAssert(trip);
  NSString *URLString = trip.plannedURLString;
  if (!URLString) {
    return;
  }
  
  [[NSUserDefaults standardUserDefaults] setObject:URLString forKey:@"TKReporterLatestPlannedURL"];

  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
    NSString *latestURLString = [[NSUserDefaults standardUserDefaults] stringForKey:@"TKReporterLatestPlannedURL"];
    if (![latestURLString isEqualToString:URLString]) {
      return; // Ignore this one
    }
    
    [SVKServer POST:[NSURL URLWithString:latestURLString]
              paras:userInfo
         completion:
     ^(id  _Nullable responseObject, NSError * _Nullable error) {
#pragma unused(responseObject)
       BOOL success = (error == nil);
       if (completeHandler) {
         completeHandler(success);
       }
       if (success) {
         [SGKLog debug:@"TKReporter" text:@"Planned trip posted successfully"];
       } else {
         [SGKLog debug:@"TKReporter" block:^NSString * _Nonnull{
           return [NSString stringWithFormat:@"Planned trip post encountered error: %@", error];
         }];
       }
     }];

  });
}

+ (void)reportProgressForTrip:(Trip *)trip
                    locations:(NSArray *)locations
{
  NSParameterAssert(trip);
  NSParameterAssert(locations);
  if (!trip.progressURLString) {
    return;
  }
  
  NSMutableArray *samples = [NSMutableArray arrayWithCapacity:locations.count];
  for (id object in locations) {
    if ([object isKindOfClass:[CLLocation class]]) {
      CLLocation *location = object;
      NSMutableDictionary *sample = [NSMutableDictionary dictionaryWithCapacity:5];
      sample[@"timestamp"] = @([location.timestamp timeIntervalSince1970]);
      sample[@"latitude"] = @(location.coordinate.latitude);
      sample[@"longitude"] = @(location.coordinate.longitude);
      if (location.speed >= 0) {
        sample[@"speed"] = @(location.speed);
      }
      if (location.course >= 0) {
        sample[@"bearing"] = @(location.course);
      }
      [samples addObject:sample];
    }
  }
  
  NSDictionary *paras = @{@"samples": samples};
  
  [SVKServer POST:[NSURL URLWithString:trip.progressURLString]
            paras:paras
       completion:
   ^(id  _Nullable responseObject, NSError * _Nullable error) {
#pragma unused(responseObject)
     if (error) {
       [SGKLog debug:@"TKReporter" block:^NSString * _Nonnull{
         return [NSString stringWithFormat:@"Progress post encountered error: %@", error];
       }];
     } else {
       [SGKLog debug:@"TKReporter" text:@"Progress posted successfully"];
     }
   }];
}

@end

File: Classes/server/TKReporter.swift
//
//  TKReporter.swift
//  TripKit
//
//  Created by Adrian Schoenig on 23/11/16.
//
//

import Foundation

import CoreLocation

public class TKReporter {
  
  private init() { }
  
  public static func reportPlannedTrip(_ trip: Trip, userInfo: [String: Any], completion: ((Bool) -> Void)?) {
    
    guard
      let urlString = trip.plannedURLString,
      let url = URL(string: urlString) else { return }
    
    let key = "TKReporterLatestPlannedURL"
    UserDefaults.standard.set(url, forKey: key)
    
    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 5) {
      
      // Only report if url is still the last one planned
      guard let currentUrl = UserDefaults.standard.url(forKey: key), currentUrl == url
        else { return }
      
      UserDefaults.standard.removeObject(forKey: key)
      
      SVKServer.post(url, paras: userInfo) { _, _, error in
        if let error = error {
          SGKLog.debug("TKReporter", text: "Planned trip encountered error: \(error)")
          completion?(false)
        } else {
          SGKLog.debug("TKReporter", text: "Planned trip reported successfully")
          completion?(true)
        }
      }
  
    }
    
  }
  
  
  public static func reportProgress(for trip: Trip, locations: [CLLocation]) {
    
    guard
      let urlString = trip.progressURLString,
      let url = URL(string: urlString) else { return }
    
    let samples = locations.map { $0.progressDict }
    let paras = [
      "samples": samples
    ]
    
    SVKServer.post(url, paras: paras) { _, _, error in
      if let error = error {
        SGKLog.debug("TKReporter", text: "Progress post encountered error: \(error)")
      } else {
        SGKLog.debug("TKReporter", text: "Progress posted successfully")
      }
    }
  }
  
}

extension CLLocation {
  
  fileprivate var progressDict: [String: Any] {
    var sample = [
      "timestamp":  timestamp.timeIntervalSince1970,
      "latitude":   coordinate.latitude,
      "longitude":  coordinate.longitude
    ]
    
    sample["speed"]   = speed  >= 0 ? speed : nil
    sample["beraing"] = course >= 0 ? course : nil
    return sample
  }
  
}

